stablehlo/dialect/StablehloOps.cpp

```diff
@@ -2305,6 +2305,32 @@ LogicalResult CompareOp::inferReturnTypeComponents(
     ValueShapeRange operands, DictionaryAttr attributes, RegionRange regions,
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
   CompareOp::Adaptor adaptor(operands, attributes, regions);
+  if (adaptor.getCompareType().has_value() &&
+      adaptor.getCompareType().value() != ComparisonType::NOTYPE) {
+    auto elementType =
+        adaptor.getLhs().getType().cast<ShapedType>().getElementType();
+    ComparisonType compareType = adaptor.getCompareType().value();
+    if (elementType.isa<FloatType>() && compareType != ComparisonType::FLOAT &&
+        compareType != ComparisonType::TOTALORDER)
+      return emitOptionalError(
+          location, "compareType must be FLOAT or TOTALORDER, but got ",
+          stringifyEnum(compareType));
+    else if (elementType.isa<ComplexType>() &&
+             compareType != ComparisonType::FLOAT)
+      return emitOptionalError(location, "compareType must be FLOAT, but got ",
+                               stringifyEnum(compareType));
+    else if (elementType.isa<IntegerType>() &&
+             elementType.isUnsignedInteger() &&
+             compareType != ComparisonType::UNSIGNED)
+      return emitOptionalError(location,
+                               "compareType must be UNSIGNED, but got ",
+                               stringifyEnum(compareType));
+    else if (elementType.isa<IntegerType>() &&
+             elementType.isSignlessInteger() &&
+             compareType != ComparisonType::SIGNED)
+      return emitOptionalError(location, "compareType must be SIGNED, but got ",
+                               stringifyEnum(compareType));
+  }
   return hlo::inferCompareOp(context, location, adaptor.getLhs(),
```

Comment:
@ghpvnist As of now I did not add the flags isFloat, isComplex, isUnsigned, isSigned because they depend on the values taken by the ComparisonType enum and we would have to makes changes if the enum were to change (which is unlikely though) but it does not seem clean to have parameters tied to values of an enum like that.

Reply:
Other option is sending a lowercase string, which I personally find cleaner IMO.

Reply:
Since we are using boolean flags, it shouldn't matter since they are taken care of in StablehloOps.cpp. The enum values shouldn't change, and if so this part of code should be fixed regardless of the approach. PS. I think the code is more readable refactored this way, at the cost of less precise error messages (but the error should still be obvious since we are talking about an enum value):
```C++
// StablehloOps.cpp
// ...
return hlo::inferCompareOp(context, location,
                      compareType.has_value(),
                      compareType == ComparisonType::NOTYPE,
                      compareType == ComparisonType::FLOAT,
                      compareType == ComparisonType::TOTALORDER,
                      compareType == ComparisonType::UNSIGNED,
                      compareType == ComparisonType::SIGNED,
                      adaptor.getLhs());

// TypeInference.cpp
LogicalResult inferCompareOp(
    MLIRContext* context, std::optional<Location>, bool compareTypeHasValue,
    bool isNoType, bool isFloat, bool isTotalOrder, bool isUnsigned, bool isSigned,
    Value lhs, SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
  // ...
  if (compareTypeHasValue && !isNoType) {
    auto elementType = lhs.getType().cast<ShapedType>().getElementType();
    if (elementType.isa<FloatType>() && !isFloat && !isTotalOrder)
      return emitOptionalError(location,
                               "compareType must be FLOAT or TOTALORDER");
    else if (elementType.isa<ComplexType>() && !isFloat)
      return emitOptionalError(location, "compareType must be FLOAT");
    else if (elementType.isa<IntegerType>() &&
             elementType.isUnsignedInteger() && !isUnsigned)
      return emitOptionalError(location, "compareType must be UNSIGNED");
    else if (elementType.isa<IntegerType>() &&
             elementType.isSignlessInteger() && !isSigned)
      return emitOptionalError(location, "compareType must be SIGNED");
  }
  // ...
  ```

Reply:
Made the change but keeping this open for now.